ls1intum/Artemis
# [Bugfix] Course/Exam Archiving should not fail when zipping non-existent files

### Checklist

### Motivation and Context


There was a flaky test testing the archival process of courses and exams that kept on failing during the `sever-tests` Github check. 

Furthermore the course archival process was also failing due to an uncaught `java.nio.file.NoSuchFileException` which occurred every time we attempt to zip a non existent file. In this specific case, for some reason the jGit garbage collection kicks in during the zipping. This causes a file `gc.log.lock` to be created and then shortly after deleted.

### Description

This pr fixed the issue by using the `Files.isReadable` check before zipping specific files. We also delete the `gc.log.lock` file before we zip the repo.

Instructor that triggers the archive function will get notified of warnings that occurred during the process via a modal window that opens up after the archival process is finished.

This pr also attempts to improve exception handling and adds missing `throws` statements.

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
4. Generate one (or more) exercise of each type
5. Participate in each exercise
6. Generate an exam with exercises of each type
7. Participate in the exam
8. Edit the course and change the end date so that it's in the past
9. Navigate to the course detail page and click on "archive course"
10. Once archiving is complete, you should be able to download the course archive
11. Download the archive. You should be able to see the submissions in the zip file
12. Now go to Bitbucket and/or Gitlab and delete a repository of a programming exercise
13. Archive the course again and you should see the popup once archiving is finished.

### Test Coverage



CourseExamExportService.java 84%
ZipFileService.java 94%
GitService.java 75%
ProgrammingExerciseExportService.java 71%
course-exam-archive-button.component.ts 93.58

### Screenshots



The modal that opens up if the archive process completed with warnings: 

![image](https://user-images.githubusercontent.com/11298674/115719502-6fafda80-a37c-11eb-8def-b3ec2dee2390.png)
